feat: handle PI modification by ID in broker

tmetzke opened the pull request:
## Description

Adds support for modifying process instance element instances by ID **in the broker**. No gateway (gRPC or REST) is extended here yet. This will happen in a follow-up to keep focus.

The following is now supported:
* Terminate instructions can contain an `elementId` to terminate all element instances at that ID in the process instance.
* Only one of `elementId` or `elementInstanceKey` can be provided. The command is rejected otherwise.
* New move instructions allow moving from a `sourceElementId` to a `targetElementId`.
  * All element instances at the source element ID are terminated (equivalent to a terminate by `elementId`).
  * For multi-instances, a new element instance is activated at the target **only for the multi-instance body**. The child instances are only terminated.
  * For all other source activity types, a new element instance is activated at the target.
  * The ancestor scope key for the target element instance can be provided via `ancestorScopeKey`, just like for activate instructions.
  * If the key of the source element instance's parent should be used as the ancestor key instead, set `useSourceParentKeyAsAncestorScope` to `true`.
  * All variable instructions are copied to the respective activation instructions.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes https://github.com/camunda/camunda/issues/41873

tmetzke committed:
feat: add move and terminate by id to engine protocol

Adds move instructions and terminate by id to the engine protocol and implementing classes.

tmetzke committed:
test: add move and terminate by id to engine protocol

Ensures move instructions and terminate by id are properly added to the engine protocol and implementing classes.

tmetzke committed:
feat: add move and terminate by id to Zeebe exporters

Adds move instructions and terminate by id to the Zeebe exporters (indices and bulk index handlers).

tmetzke committed:
feat: handle move and terminate by id in Zeebe

Handles move instructions and terminate by id in the command processor in Zeebe.
The instructions are resolved and added to the activate and terminate instructions by key.

tmetzke committed:
test: handle move and terminate by id in Zeebe

Verifies move instructions and terminate by id are handled as expected in the command processor in Zeebe.

tmetzke committed:
feat: use moved move instructions class in services

tmetzke committed:
test: use moved move instructions class in services

tmetzke committed:
feat: use moved move instructions class in REST

tmetzke committed:
test: use moved move instructions class in REST

tmetzke committed:
fix: use correct protocol class in REST

tmetzke commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java:
```diff
@@ -342,20 +369,116 @@ public ProcessingError tryHandleError(
+            .collect(
+                Collectors.toMap(
+                    ProcessInstanceModificationMoveInstructionValue::getSourceElementId,
+                    Function.identity()));
```
Yep, I'll add this 👍

tmetzke committed:
feat: expand terminate and move instructions validation

Adds further validation to move and terminate instructions to avoid running into unexpected behavior.

tmetzke committed:
test: expand terminate and move instructions validation

Verifies further validation of move and terminate instructions to avoid running into unexpected behavior.

tmetzke committed:
feat: support variable instructions on move modification

Adds support for variable instructions to the move instruction. They are copied to the respective activation instruction.

tmetzke committed:
test: add variable instructions on move to golden files

mustafadagher commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java:
```diff
@@ -344,20 +371,116 @@ public ProcessingError tryHandleError(
+          final long flowScopeKey = elementInstance.getValue().getFlowScopeKey();
+          if (flowScopeKey > 0
+              && !BpmnElementType.MULTI_INSTANCE_BODY.equals(
+                  elementInstanceState.getInstance(flowScopeKey).getValue().getBpmnElementType())) {
```
❓ did we have this limitation before? how did it come up?

tmetzke replied:
Done with https://github.com/camunda/camunda/pull/41649/commits/1cef42b8abd4bff67e1849595811938c9acf54a7

tmetzke replied:
Thanks for the question and making me think about it again 🙏

tmetzke replied:
Ugh, doesn't work with https://github.com/camunda/camunda/commit/1cef42b8abd4bff67e1849595811938c9acf54a7. I cannot wholly ignore the MI children, of course 😅 I can only do this when the MI body element ID is in the move or terminate-by-id instructions.

mustafadagher commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTerminationTest.java:
```diff
@@ -339,6 +342,60 @@ public void shouldTerminateAllElementsInRootScope() {
+        .withInstanceKey(processInstanceKey)
+        .modification()
+        .activateElement("A")
+        .modify();
```
I see you already have such a test case below

tmetzke committed:
refactor: handle MI children implicitly

Multi-instance children don't need explicit terminate instructions for move/cancel by ID.
The engine terminates them already when the parent MI body is terminated.
We don't want any activation for the children either.

tmetzke committed:
fix: improve terminate instruction rejection message

Clarify that only one of `elementId` and `elementInstanceKey` is allowed for terminations.

tmetzke commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceRejectionTest.java:
```diff
@@ -1342,4 +1343,46 @@ public void shouldAcceptActivationWithValidAncestorScopeKeyInEventSubProcess() {
+        .describedAs("Expect that activation is rejected when flow scope is terminated")
+        .hasRejectionType(RejectionType.INVALID_ARGUMENT)
+        .hasRejectionReason(
+            ("Expected to modify instance of process '%s' but it contains one or more terminate instructions "
```
Done with https://github.com/camunda/camunda/pull/41649/commits/f522c9417cba49321b9e427dbda10c2e263e29af

tmetzke committed:
test: use model API instead of file

tmetzke committed:
refactor: handle child elements correctly in move

If an element is terminated, the child elements are not handled for ID-based instructions anymore.

mustafadagher reviewed (approved):
Thanks a lot @tmetzke ! 🚀 
The PR is kind of huge, yet brings great improvements and it's well covered by tests. LGTM!

mustafadagher commented:
🔧 @tmetzke an after thought:
I created the tests here(https://github.com/camunda/camunda/pull/40453) to verify certain use-cases needed for https://github.com/camunda/camunda/issues/40179 are supported by the engine.

💭 Could we replicate all those tests and make sure they pass using a `move` instruction instead of a terminate+activate ones?

We can also choose to take that on in a follow-up PR if you feel strongly it doesn't fit here.

tmetzke committed:
test: extend move instruction tests

Adds test cases for the following move scenarios:
- Move into a specific scope by key (multiple instances of the scope exist)
- Move inside multiple ad-hoc sub-process instances
- Move inside multiple nested ad-hoc sub-process instances

tmetzke commented:
> 🔧 @tmetzke an after thought: I created the tests here(#40453) to verify certain use-cases needed for #40179 are supported by the engine.
> 
> 💭 Could we replicate all those tests and make sure they pass using a `move` instruction instead of a terminate+activate ones?
> 
> We can also choose to take that on in a follow-up PR if you feel strongly it doesn't fit here.

I added equivalent move tests for the first two of your test cases.
I also added a dedicated test for the need for an ancestor scope key for move.
I believe your third test case requires `sourceElementInstanceKey` for move instructions, which we don't have yet here in this PR.
